File: Frontend/reader_frontend_app/lib/models/sale_trade.dart
class SaleTrade {
  final int id;
  final int idUser;
  final String isbn;
  final String title;
  final String? desiredBookTitle;
  final String username; // Novo campo para o nome do vendedor
  final double? price;
  final bool isAvailableForSale;
  final bool isAvailableForTrade;
  final String state;
  final String? notes;
  final String? isbnDesiredBook;
  final DateTime dateCreation;

  SaleTrade({
    required this.id,
    required this.idUser,
    required this.isbn,
    required this.title,
    this.price,
    this.desiredBookTitle,
    required this.isAvailableForSale,
    required this.isAvailableForTrade,
    required this.state,
    this.notes,
    this.isbnDesiredBook,
    required this.username,
    required this.dateCreation,
  });

  factory SaleTrade.fromJson(Map<String, dynamic> json) {
    return SaleTrade(
      id: json['id'] ?? (throw Exception('Campo `id` é nulo')),
      idUser: json['idUser'] ?? (throw Exception('Campo `idUser` é nulo')),
      isbn: json['isbn'] ?? 'ISBN não disponível',
      title: json['title'] ?? 'Título não disponível',
      price: json['price'] != null ? (json['price'] as num).toDouble() : null,
      desiredBookTitle:
      json['desiredBookTitle'] ?? 'Título do livro desejado não disponível',
      isAvailableForSale: json['isAvailableForSale'] ?? false,
      isAvailableForTrade: json['isAvailableForTrade'] ?? false,
      state: json['state'] ?? 'Indefinido',
      username: json['username'] ?? 'Nome do vendedor não disponível',
      notes: json['notes'],
      isbnDesiredBook: json['isbnDesiredBook'],
      dateCreation: json['dateCreation'] != null
          ? DateTime.parse(json['dateCreation'])
          : DateTime.now(),
    );
  }
}

File: Backend/Read.er/Models/SaleTrades/SaleTrade.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using Read.er.Enumeracoes;

namespace Read.er.Models.SaleTrades;

/// <summary>
/// Represents a sale or trade offer for a book in the system.
/// This class is used to manage the details and availability of a book for sale or trade by a user.
/// </summary>
public class SaleTrade
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }

    [Required]
    public int IdUser { get; set; }

    [Required]
    [MaxLength(13)]
    [RegularExpression("(.{10}|.{13})", ErrorMessage = "Comprimento deve ser 10 ou 13.")]
    public string Isbn { get; set; } = string.Empty;

    public decimal? Price { get; set; }

    public bool IsAvailableForSale { get; set; }
    public bool IsAvailableForTrade { get; set; }

    [Required]
    public BookState State { get; set; }

    [MaxLength(500, ErrorMessage = "O comprimento não pode exceder os 500 caracteres.")]
    public string? Notes { get; set; }

    [MaxLength(13)]
    [RegularExpression("(.{10}|.{13})", ErrorMessage = "Comprimento deve ser 10 ou 13.")]
    public string? IsbnDesiredBook { get; set; }

    [Required]
    public DateTime DateCreation { get; set; } = DateTime.UtcNow;

    // Adicionar título do livro
    [MaxLength(80, ErrorMessage = "O comprimento não pode exceder os 80 caracteres.")]
    public string Title { get; set; } = string.Empty;

    public string? DesiredBookTitle { get; set; } = string.Empty;


    [JsonIgnore]
    public ICollection<SaleTradeOffer> Offers { get; set; } = new List<SaleTradeOffer>();
}
